File: 0073-set-matrix-zeroes/solution.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        rows, cols = len(matrix), len(matrix[0])
        marked = [[0] * cols for _ in range(rows)]  # Proper initialization
        
        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == 0 and marked[i][j] != 1:
                    # Mark entire row and column
                    for k in range(cols):
                        if matrix[i][k] != 0:
                            matrix[i][k] = 0
                            marked[i][k] = 1
                    for k in range(rows):
                        if matrix[k][j] != 0:
                            matrix[k][j] = 0
                            marked[k][j] = 1

File: 0073-set-matrix-zeroes/solution.java
class Solution {
    public void setZeroes(int[][] matrix) {
        Set<Integer> rowSet = new HashSet<>();
        Set<Integer> colSet = new HashSet<>();

        List<int[]> list = new LinkedList<>();

        for(int i = 0; i < matrix.length; i++) {
            for(int j = 0; j < matrix[0].length; j++) {
                if(matrix[i][j] == 0) {
                    list.add(new int[]{i, j});
                }
            }
        }

        for(int[] i : list) {
            setZero(matrix, i[0], i[1], rowSet, colSet);
        }

    }

    private void setZero(int[][] matrix, int row, int col, Set<Integer> rowSet, Set<Integer> colSet) {
        if(!colSet.contains(col)) {
            for(int i = 0; i < matrix.length; i++) {
                matrix[i][col] = 0;
            }
            colSet.add(col);
        }
        if(!rowSet.contains(row)) {
            for(int i = 0; i < matrix[0].length; i++) {
                matrix[row][i] = 0;
            }
            rowSet.add(row);
        }
    }
}
